### PR TITLE
Charlesmchen/fix attachment uploads

### PR DESCRIPTION
**Pull request by charlesmchen**
This PR is now ready for review.

PTAL @michaelkirk

### CODE REVIEW

**Comment by charlesmchen** [comment]
This PR is now ready for review.

PTAL @michaelkirk

**Comment by charlesmchen** [comment]
@michaelkirk Hmm, layout is broken.  Investigating.

**Comment by charlesmchen** [comment]
Okay, fixed this issue.

File: Signal/src/ViewControllers/ConversationView/ConversationInputToolbar.m [hunk 1]
```diff
@@ -344,7 +344,7 @@ - (void)ensureContentConstraints
 
 - (void)ensureShouldShowVoiceMemoButton
 {
-    self.shouldShowVoiceMemoButton = (self.attachmentToApprove != nil && self.inputTextView.trimmedText.length < 1);
+    self.shouldShowVoiceMemoButton = (self.attachmentToApprove == nil && self.inputTextView.trimmedText.length < 1);
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationInputToolbar.m]: The bug you noted in the last PR.

File: Signal/src/ViewControllers/ConversationView/Cells/AttachmentUploadView.m [hunk 2]
```diff
@@ -131,6 +128,45 @@ - (void)attachmentUploadProgress:(NSNotification *)notification
...
+    [super setFrame:frame];
+    if (sizeDidChange) {
+        [self updateLayout];
+    }
+}
+
+- (void)updateLayout
+{
+    // Center the progress bar within the bubble mask.
+    //
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/Cells/AttachmentUploadView.m]: It's easier to express the layout of the progress bar using manual layout.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m [hunk 3]
```diff
@@ -706,11 +712,14 @@ - (void)addAttachmentUploadViewIfNecessary:(UIView *)attachmentView
 {
     OWSAssert(attachmentView);
     OWSAssert(attachmentStateCallback);
+    OWSAssert(self.attachmentStream);
 
     if (self.isOutgoing) {
-        self.attachmentUploadView = [[AttachmentUploadView alloc] initWithAttachment:self.attachmentStream
-                                                                           superview:attachmentView
-                                                             attachmentStateCallback:attachmentStateCallback];
+        if (!self.attachmentStream.isUploaded) {
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: Small optimization - don't both with attachment upload view if we don't need it.

Comment (michaelkirk) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: Will this show for failed uploads which aren't "in progress"?

Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: Nope, there's logic in the upload view to hide in that case.

**Pull request merged**